File: src/test/java/org/example/eventhub/service/TicketServiceTest.java
package org.example.eventhub.service;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.math.BigDecimal;
import org.example.eventhub.entity.Event;
import org.example.eventhub.entity.Order;
import org.example.eventhub.entity.Ticket;
import org.example.eventhub.entity.User;
import org.example.eventhub.exception.NoAvailableTicketsException;
import org.example.eventhub.repository.TicketRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class TicketServiceTest {

    private static final Long EVENT_ID = 100L;
    private static final Long USER_ID = 10L;
    private static final Long ORDER_ID = 1L;

    private final TicketRepository repository = mock(TicketRepository.class);
    private final UserService userService = mock(UserService.class);
    private final EventService eventService = mock(EventService.class);

    private final TicketService ticketService = new TicketService(repository, userService, eventService);

    private User user;
    private Event event;
    private Order order;

    @BeforeEach
    void setUp() {
        user = new User();
        user.setId(USER_ID);

        event = new Event();
        event.setId(EVENT_ID);
        event.setCapacity(1);
        event.setPrice(BigDecimal.valueOf(2500));

        order = Order.builder().id(ORDER_ID).build();
    }

    @Test
    @DisplayName("createTicket: успешно создаёт билет, если есть свободные места")
    void createTicket_success_whenTicketsAvailable() {
        Ticket input = Ticket.builder()
            .event(event)
            .user(user)
            .order(order)
            .price(event.getPrice())
            .build();

        when(userService.getUserByIdAsEntity(USER_ID)).thenReturn(user);
        when(eventService.getEventByIdAsEntity(EVENT_ID)).thenReturn(event);
        doNothing().when(eventService).saveEvent(event);
        when(repository.save(any())).thenReturn(input);


        Ticket result = ticketService.createTicket(EVENT_ID, USER_ID, order);

        assertNotNull(result);
        assertEquals(order, result.getOrder());
        assertEquals(event, result.getEvent());
        assertEquals(user, result.getUser());
        assertEquals(event.getPrice(), result.getPrice());

        verify(userService).getUserByIdAsEntity(USER_ID);
        verify(eventService).getEventByIdAsEntity(EVENT_ID);
    }

    @Test
    @DisplayName("createTicket: бросает NoAvailableTicketsException, если билетов нет")
    void createTicket_throwsException_whenNoTicketsAvailable() {
        Event eventWithoutTickets = Event.builder()
                .id(EVENT_ID)
                .capacity(0)
                .build();

        when(userService.getUserByIdAsEntity(USER_ID)).thenReturn(user);
        when(eventService.getEventByIdAsEntity(EVENT_ID)).thenReturn(eventWithoutTickets);

       assertThrows(NoAvailableTicketsException.class, () -> ticketService.createTicket(EVENT_ID, USER_ID, order));


        verify(userService).getUserByIdAsEntity(USER_ID);
        verify(eventService).getEventByIdAsEntity(EVENT_ID);
    }
}

File: src/test/kotlin/org/example/eventhub/service/TicketServiceTest.kt
package org.example.eventhub.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.example.eventhub.entity.Event
import org.example.eventhub.entity.Order
import org.example.eventhub.entity.Ticket
import org.example.eventhub.entity.User
import org.example.eventhub.enums.OrderStatus
import org.example.eventhub.enums.Role
import org.example.eventhub.exception.NoAvailableTicketsException
import org.example.eventhub.repository.TicketRepository
import java.math.BigDecimal
import java.time.LocalDateTime

class TicketServiceTest : StringSpec({

    val repository = mockk<TicketRepository>()
    val eventService = mockk<EventService>()

    val service = TicketService(repository, eventService)

    val user = User(
        id = 1,
        username = "username",
        email = "email@email.com",
        password = "password",
        role = Role.USER
    )
    val order = Order(
        id = 1,
        user = user,
        totalPrice = BigDecimal.ZERO,
        createdAt = LocalDateTime.of(2021, 1, 1, 1, 1, 0),
        status = OrderStatus.PENDING,
        reservedUntil = LocalDateTime.of(2021, 1, 1, 1, 16, 0)
    )

    "createTicket создает билет, учитывает это в event, сохраняет в репозиторий" {
        val event = Event(
            id = 1,
            title = "title",
            dateTime = LocalDateTime.of(2021, 1, 1, 1, 1, 0),
            capacity = 1,
            organizer = user
        )
        val ticket = Ticket(
            id = 1,
            event = event,
            user = user,
            order = order,
            price = BigDecimal.ZERO,
        )

        every { eventService.getEventByIdAsEntity(any()) } returns event
        every { eventService.saveEvent(any()) } returns event
        every { repository.save(any()) } returns ticket

        service.createTicket(1, user, order)

        verify(exactly = 1) { repository.save(any()) }
        verify(exactly = 1) { eventService.saveEvent(any()) }
        verify(exactly = 1) { eventService.getEventByIdAsEntity(any()) }
    }

    "createTicket выбрасывает ошибку, если билетов на мероприятие больше нет" {
        val event = Event(
            id = 1,
            title = "title",
            dateTime = LocalDateTime.of(2021, 1, 1, 1, 1, 0),
            capacity = 0,
            organizer = user
        )

        every { eventService.getEventByIdAsEntity(any()) } returns event

        val ex = shouldThrow<NoAvailableTicketsException> {
            service.createTicket(1, user, order)
        }

        ex.message shouldBe "Свободных билетов для ${event.title} не осталось"
    }
})